messages/src/lib.rs

```diff
@@ -89,6 +101,80 @@ impl std::fmt::Display for Error {
     }
 }
 
+#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Eq, Serialize, SerializedSize)]
+#[cfg_attr(any(test, feature = "std"), derive(thiserror::Error))]
+pub enum HwError {
+    /// Could not set the reset pin
+    SetResetFailed,
+
+    /// Could not clear the reset pin
+    ClearResetFailed,
+
+    /// Failed to clear the power enable mask
+    ClearPowerEnableFailed,
+
+    /// Failed to set the power enable mask
+    SetPowerEnableFailed,
+
+    /// Failed to clear the low power mode mask
+    ClearLpModeFailed,
+
+    /// Failed to set the low power mode mask
+    SetLpModeFailed,
+
+    /// Failed to read the `EN` register
+    EnableReadFailed,
+
+    /// Failed to read the `RESET` register
+    ResetReadFailed,
+
+    /// Failed to read the `LPMODE` register
+    LpReadFailed,
+
+    /// Failed to read the `PRESENT` register
+    PresentReadFailed,
+
+    /// Failed to read the `IRQ` register
+    IrqReadFailed,
+
+    /// Could not set up the write buffer for a page select
+    PageSelectWriteBufFailed,
+
+    /// The page select write operation failed
+    PageSelectWriteFailed,
+
+    /// Could not set up the write buffer for a bank select
+    BankSelectWriteBufFailed,
+
+    /// The bank select write operation failed
+    BankSelectWriteFailed,
+
+    /// Waiting for the operation to complete failed
+    WaitFailed,
+
+    /// The FPGA reported an I2C error
+    I2cError,
+
+    /// The read setup operation failed
+    ReadSetupFailed,
+
+    /// Reading back the read buffer failed
+    ReadBufFailed,
+
+    /// The write setup operation failed
+    WriteSetupFailed,
+
+    /// The write operation failed
+    WriteFailed,
```

Comment:
Is this a catch-all for a generic write failure? Or does this mean something specific in terms of the SP <-> FPGA interface?